docs/upgrade.html

```diff
@@ -139,7 +139,9 @@ <h5><a id="upgrade_420_notable" href="#upgrade_420_notable">Notable changes in 4
                 <ul>
                     <li>Null values are no longer accepted for most LIST-type configurations, except those that explicitly
                         allow a null default value or where a null value has a well-defined semantic meaning.</li>
-                    <li>Duplicate entries within the same list are no longer permitted.</li>
+                    <li>Most LIST-type configurations no longer accept duplicate entries, except in cases where duplicates
```

Comment:
1. Should we change ValidList.ensureValid() to only log a warning for duplicated values?
2. The following makes the generated "Valid Values" in html quite verbose. It would be useful to improve that. For example, if empty is not allowed, we probably don't need to say it. If empty is allowed, we could add "empty".
```
        public String toString() {
            return validString + (isEmptyAllowed ? " (empty config allowed)" : " (empty not allowed)") +
                    (isNullAllowed ? " (null config allowed)" : " (null not allowed)");
        }
```
Also, "Valid Values" for `anyNonDuplicateValues` will show up as `[]`, which is counter intuitive.

Reply:
> Should we change ValidList.ensureValid() to only log a warning for duplicated values?

I think we can keep the exception-throwing behavior, but we now preprocess the user’s configuration to remove duplicate entries.

> The following makes the generated "Valid Values" in html quite verbose. 

I have updated the document
<img width="645" height="355" alt="asdfsfd" src="https://github.com/user-attachments/assets/8402a0e3-53aa-46ae-9cf2-2c3cd2be8ab0" />
<img width="631" height="198" alt="sdaf" src="https://github.com/user-attachments/assets/128a633a-50a5-4406-89ae-5b67defa35db" />
<img width="636" height="279" alt="dlfsakfds" src="https://github.com/user-attachments/assets/8f605f42-3706-45f4-af3b-71b6a7ec90f6" />

Reply:
I have updated the document:

<img width="676" height="340" alt="CleanShot 2025-11-19 at 19 25 38" src="https://github.com/user-attachments/assets/5d2242ba-c59a-4780-adf2-2b510680e99e" />
<img width="637" height="279" alt="CleanShot 2025-11-19 at 19 26 26" src="https://github.com/user-attachments/assets/4a529e0e-1799-4633-b90c-52f9738b4774" />
<img width="644" height="267" alt="CleanShot 2025-11-19 at 19 26 54" src="https://github.com/user-attachments/assets/9d490a4a-75c8-44a4-9996-d9885d0e7dd4" />